modules/ipp/scanner.go

```diff
@@ -126,86 +172,394 @@ func (scanner *Scanner) GetPort() uint {
+	scan.results.Response = resp
 	if resp != nil && resp.Body != nil {
 		defer resp.Body.Close()
 	} else {
-		// FIXME: Is empty body allowed in IPP?
-		// Cite RFC!!
-		// Empty body is not allowed in valid IPP
-		return 0, 0, zgrab2.NewScanError(zgrab2.SCAN_PROTOCOL_ERROR, nil)
+		// resp == nil or resp.Body == nil
+		// Empty response/body is not allowed in IPP because a response has required parameter
+		// Source: RFC 8011 Section 4.1.1 https://tools.ietf.org/html/rfc8011#section-4.1.1
+		// Still returns the response, if any, because assignment occurs before this else block
+		// TODO: Examine whether an empty response overall is a protocol error, I'd think of it as another kind of error entirely,
+		//       and later conditions might handle that case; see RFC 8011 Section 4.2.5.2?
+		return zgrab2.NewScanError(zgrab2.SCAN_PROTOCOL_ERROR, nil)
 	}
+	if err != nil {
+		//If error is a url.Error (a struct), unwrap it
+		if urlError, ok := err.(*url.Error); ok {
+			err = urlError.Err
+		}
+	}
+	if err != nil {
+		switch err {
+		case ErrRedirLocalhost:
+			break
+		case ErrTooManyRedirects:
+			return zgrab2.NewScanError(zgrab2.SCAN_APPLICATION_ERROR, err)
+		default:
+			return zgrab2.DetectScanError(err)
+		}
+	}
+	storeBody(resp, scanner)
 
-	// FIXME: Maybe add something to handle redirects
-	// FIXME: Probably return the whole response for further inspection by ztag, rather
-	//         than grabbing first 2 bytes. In that case, implement maxRead like http module
-
-	//Check to make sure that the repsonse received is actually IPP
+	// TODO: Check to make sure that the repsonse received is actually IPP
 	//Content-Type header matches is sufficient
 	//HTTP on port 631 is sufficient
 	//Still record data in the case of protocol error to see what that data looks like
 
-	// TODO: Record server-header version numbers
-	//protocols := resp.Header.Get("Server")
-	// TODO: Change this to perform two separate binary.Reads
-	var version uint16
-	// TODO: Determine whether errors other than protocol (ie: too few bytes) can be triggered here
-	if err := binary.Read(resp.Body, binary.BigEndian, &version); err != nil {
-		// FIXME: Determine whether sending fewer than 2 bytes is a protocol or application error
-		// I believe it's protocol, since the version must be specified (iirc)
-		// FIXME: Cite RFC!!
-		return 0, 0, zgrab2.NewScanError(zgrab2.SCAN_PROTOCOL_ERROR, err)
+	protocols := strings.Split(resp.Header.Get("Server"), " ")
+	for _, p := range protocols {
+		if strings.HasPrefix(strings.ToUpper(p), "IPP/") {
+			scan.results.VersionString = p
+			protocol := strings.Split(p, "/")[1]
+			components := strings.Split(protocol, ".")
+			// Reads in signed integers because "every integer MUST be encoded as a signed integer"
+			// (Source: https://tools.ietf.org/html/rfc8010#section-3)
+			var major, minor int8
+			if len(components) >= 1 {
+				if val, err := strconv.Atoi(components[0]); err != nil {
+					// TODO: Handle error
```

Comment:
I'd say for now, add a debug-level log, with a TODO to remove it -- that could help debug any issues we in the 1% scan.

That would go for any other places where there are potentially errors that you can't do anything about, but indicate unexpected behavior.